### ISSUE TITLE
[css-view-transitions-1] Clarify implications of scrolling during a transition

### ISSUE THREAD

**Issue by mattwoodrow** [w3c/csswg-drafts#10197]
My understanding is that the computed value for 'transform' that is set for the `::view-transition-group` selector includes the current scroll position of the Document (as required to place it in the current visual position if the view-transition is started when scrolled down).

This means that the `-ua-view-transition-group-anim-transitionName` 'from' keyframe includes the initial scroll position, and the (implicit) 'to' keyframe includes the current scroll position (and changes each frame if scrolling happens).

This has the visual effect that scrolling lags behind the movement of the scrollbar (and feels heavy), since the pseudo elements animate to the new scroll position. Obviously this is more pronounced when using a longer transition duration.

Assuming that this is the intended behaviour, it might be worth adding a non-normative explainer so that it's more obvious that this will happen.

**Comment by vmpstr** [comment]
The behavior of targetting the new location of the element is, of course, intentional, since we want to retarget to the new position anytime that changes due to scroll or any other change.

The behavior of the old position being fixed is a consequence of the fact that we need to support cases where the "old" element is gone. It might have been affected by the scroll, or might have been in a different scroller, or be fixedpos. I would claim that this is intentional.

@noamr I'm not sure what the right spot to add a non-normative note is, but possibly around the capture steps?

**Comment by khushalsagar** [comment]
+1 that this is working as intended. We retarget the end state of the group's animation to the new element's current state including all sources which contribute to the element's position.

But our implicit assumption was that interactions like scrolling won't happen during the transition since its a transient state for animations and the pseudo-elements consume all the events (via the default UA CSS). If that doesn't work for a use-case, I'd be happy to hear the details and see if we can do anything better.

**Comment by bramus** [comment]
Digging this back up after I noticed that Firefox (currently) doesn’t account for the adjusted scroll offset (https://bugzilla.mozilla.org/show_bug.cgi?id=1991967) during a View Transition.

(This was most likely overlooked, because the details in this issue have not made it into the spec yet 😬)

I read in this issue here that:

> the `-ua-view-transition-group-anim-transitionName` 'from' keyframe includes the initial scroll position, and the (implicit) 'to' keyframe includes the current scroll position (and changes each frame if scrolling happens).

To me this seems like a lot of work, especially when you have a lot of groups.

Wouldn’t it be more performant if the entire `::view-transition` pseudo moved as a whole, instead of updating all end positions of each and every `::view-transition-group()`?

This approach is exactly what I do for the Firefox workaround in https://codepen.io/bramus/pen/KwVMMdr/510ccfa2f1599ac75233d78eb0376333: while a View Transition is running I track the scroll delta and update the `::view-transition`’s position accordingly.

```css
::view-transition {
  translate: calc(var(--vt-scroll-delta-x, 0) * 1px) calc(var(--vt-scroll-delta-y, 0) * 1px);
}
```

I believe this would also play nice with https://github.com/w3c/csswg-drafts/issues/12324, in which there is discussion about how to position the `::view-transition` pseudo for scoped transitions, maybe using anchor positioning.

**Comment by bramus** [comment]
https://codesandbox.io/p/sandbox/m7zyzc _(forked from a demo by @joshwcomeau)_ nicely demonstrates the problem with the current specced approach.

Recording:

https://github.com/user-attachments/assets/160a3ac5-ec66-4b20-a404-e337249f9ae6

With the long `animation-duration` used in the demo, the problem becomes very much visible: because the target scroll position gets updated in the target value of the subject, scrolling doesn’t happen immediately but is controlled by the `animation-duration`.

---

https://codesandbox.io/p/sandbox/xldyq8 is the adjusted version of that demo, that uses the approach as described above: it updates the position of the `::view-transition` on scroll. The result is that it correctly responds to scroll.

Recording:

https://github.com/user-attachments/assets/37f32967-ff90-41a8-882d-89b5a7ca2d39

**Comment by vmpstr** [comment]
`::view-transition` uses the snapshot containing block which is not affected by the document scroll, so having us move the view transition pseudo in response to the scroll would change that semantic. Conceptually the document view transition happens "above" the document, so everything is re-targeted in response to the scroll instead of adjusting immediately.

I can definitely see the argument for keeping this "internal" to the scroller, but to explain my position consider the following case where the scroller is not the root scroller (https://vt-retarget-vmpstr.yoyo.codes/):

https://github.com/user-attachments/assets/b7445eb5-1939-46f4-8530-34fae285f1d4

Here, when the ball remains within the scroller it still retargets, but it should be clear it's because the transition is running outside of the scroller. Is this the intended effect? I don't know, maybe the participating element should be moving within the scroller responding instantly. Although escaping the clip would be weird in that case.

But I hope we can agree, when the ball toggles between being inside and outside, retargeting seems to be the right behavior.

My contentious claim is that the root scroller is no different here, since the transition is happening outside of it and could, for example, be transitioning from/to the top layer (which iirc isn't affected by root scroll?).

I'm sympathetic that there should be some solution to this. I think the element scoped view transitions provides a sufficiently granular tool in that it allows the transition to happen on some subtree, with the `::view-transition` pseudo hanging off that subtree's root, meaning any ancestor scrolls (like document scrolls) would instantaneously take effect.

While we work on that, I think we may want to be stick with the css/script math that @bramus has.

**Comment by bramus** [comment]
I think I’ve found a simpler solution that still retargets the end position of the `::view-transition-group(*)` pseudos but in such a way that it’s not affected by the animation duration and so that it also leaves the `::view-transition` untouched (as per @vmpstr’s remark): a `translate` that tracks the scroll delta, applied onto the `::view-transition-group`s.

Three demos:

1. The problem: https://codepen.io/bramus/pen/QwEeOVo/15db2ed6632747c69b0a75b8471e30c6 (Don't use Firefox though, see https://bugzilla.mozilla.org/show_bug.cgi?id=1991967)
2. The fix, by manipulating the `::view-transition` overlay: https://codepen.io/bramus/pen/yyJmPQm/7c91e973ecfa235d4861bc64a8d5a713
3. The revisited fix, by manipulating the `::view-transition-group(*)` pseudos: https://codepen.io/bramus/pen/PwzMEPx/c7cc82968abb00a100057c903fcd3f1b

I believe this would also play nice with elements that are inside a scroller, as the `::view-transition` overlay remains untouched. For that to work, it would need to track the `scrollParent` of the source element instead of blindly tracking the `documentElement`’s scrollOffsets.

Question that goes along with this alt fix, is whether that is web compatible or not, as it can clash with sites that are using `translate` in their own keyframes. An alternative here is to set the `translate` on the `::view-transition-image-pair(*)` _(but that might seem a bit weird)_ or to add an extra pseudo whose sole responsibility is to have it respond to scroll.

**Comment by jakearchibald** [comment]
@bramus what would happen for elements inside multiple scrollers, would look at all scroll parents?

**Comment by bramus** [comment]
Yes. In order to have the best visual effect it should track the scroll offsets of all ancestor scrollers of the element (in the new state).

Quick fork of Vlad’s demo that does this: https://codepen.io/bramus/pen/VYKZxpy/dbc3f168b9d4b96df1aba216ed70998c

- When you hit the “Toggle Inside” button, you can scroll both the inner and outer (`document`) scroller, with the dot updating accordingly.
- When you hit the “Toggle Outside” button, only the root scroller affects things, as the dot is no longer contained within `#scroller`.
- When you hit the “Toggle Outside” button again, both scrollers affect the position of the dot again, as the dot is now a child of the `#scroller` again.

**Comment by nickcoury** [comment]
We've also run into this. You can see an example here on a cross-doc VT (slowed down for emphasis) - https://photos.app.goo.gl/CdcQVUyyYKcuoiheA

I understand the reasons for anchoring to the viewport, but anchoring to scrolling seems like it would be the right default behavior more often than not given the new pseudos are live and will shift the majority of the viewport content. The only content in the old pseudo snapshots that wouldn't expect to move are sticky elements, and that makes up the minority of the visible content.

The prevalence of this should be reduced once Scoped View Transitions are available, though in theory the same issue could arise in scrollers within the scoped content.

It feels like a net improvement in smoothness and net reduction in jank change the default behavior to track scrolling, especially if it can be done on the compositor to avoid frame lags. Given the existing behavior is already very broken looking, I wouldn't be overly concerned about changing the behavior. If absolutely required, it could be an opt-in to decide whether to track scrolling and perhaps how to do the scroll tracking if multiple approaches are needed (root vs nested scroller).
